protocol/deploys/helpers.ts

```diff
@@ -0,0 +1,94 @@
+import { network } from 'hardhat';
+import {
+  BaseContract,
+  BigNumber,
+  ContractFactory,
+  ContractTransaction,
+} from 'ethers';
+import fs from 'fs';
+
+/**
+ * Wait for a transaction to mine
+ */
+export async function mine(tx: Promise<ContractTransaction>) {
+  await (await tx).wait();
+}
+
+/**
+ * Typesafe helper that works on contract factories to create, deploy, wait till deploy completes
+ * and output useful commands to setup etherscan with contract code
+ */
+export async function deployAndMine<
+  T extends BaseContract,
+  D extends (...args: any[]) => Promise<T>
+>(
+  name: string,
+  factory: ContractFactory,
+  deploy: D,
+  ...args: Parameters<D>
+): Promise<T> {
+  if (factory.deploy !== deploy) {
+    throw new Error("Contract factory and deploy method don't match");
+  }
+
+  const verifyRenderedArgs: any[] = [];
+  for (const arg of args) {
+    if (arg instanceof Buffer) {
+      verifyRenderedArgs.push(`0x${arg.toString('hex')}`)
+    } else if (arg instanceof BigNumber) {
+      verifyRenderedArgs.push(arg.toString())
+    } else {
+      // XXX(jeeva): As we deploy smart contracts, we'll realise which types we need to fix here
+      verifyRenderedArgs.push(arg);
+    }
+  }
+
+  console.log(`*******Deploying ${name} on ${network.name} with args ${JSON.stringify(verifyRenderedArgs, null, 2)}`);
+  const contract = await factory.deploy(...args) as T;
+  console.log(`Deployed... waiting for transaction to mine`);
+  console.log();
+  await contract.deployed();
+  console.log('Contract deployed');
+  console.log(`export ${name}=${contract.address}`);
+
+  const deployArgsDir = `${__dirname}/${network.name}`;
+  if (!fs.existsSync(deployArgsDir)) {
+    fs.mkdirSync(deployArgsDir);
+  }
+  fs.writeFileSync(`${deployArgsDir}/${contract.address}.js`, `module.exports = ${JSON.stringify(verifyRenderedArgs, null, 2)}`)
+
+  console.log(`hardhat verify args written to ${deployArgsDir}/${contract.address}.js`);
```

Comment:
A good step in the right direction for sure! The flow still means that after each contract deploy, they have to go in and update `contract-addresses.ts` manually, one by one.

Think we could enhance by:
1. Pass through the target of this deploy as an arg (rather than just some rando `name` string) eg `TEMPLE_V2_FRAX_PAIR`
2. After it's deployed, update the `contract-addresses.ts` `DEPLOYED_CONTRACTS.${network.name}.${target}`

Less copy/paste messup risk then.

After this is out, I think we can then automate the verify step which would be awesome too (or at least bulk verify?)